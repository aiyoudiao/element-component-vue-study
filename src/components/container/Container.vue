<script setup lang="ts">
import type { Component, VNode } from 'vue'
import { computed, toRefs, useSlots, withDefaults } from 'vue'
import { DirectionEnum } from './DirectionEnum'
const props = withDefaults(defineProps<ContainerType>(), {
  direction: DirectionEnum.horizontal,
})

interface ContainerType {
  direction?: DirectionEnum.horizontal | DirectionEnum.vertical
}
function useClasses(props) {
  const { direction } = toRefs<ContainerType>(props)
  const slots = useSlots()

  return computed(() => {
    let isVertical = false
    if (slots && slots.default) {
      const vNodes: Array<VNode> = slots.default()
      isVertical = vNodes.some((vNode) => {
        const component = vNode.type as Component
        return ['ElHeader', 'ElFooter'].includes(component.name)
      })
    }

    if (isVertical)
      return 'is-vertical'

    if (direction.value === DirectionEnum.vertical)
      return 'is-vertical'
    if (direction.value === DirectionEnum.horizontal)
      return ''

    return ''
  })
}

const classes = useClasses(props)
</script>

<template>
  <section class="el-container" :class="classes">
    <slot />
  </section>
</template>

<style lang="scss">
@import '../../theme/src/container.scss';
</style>
