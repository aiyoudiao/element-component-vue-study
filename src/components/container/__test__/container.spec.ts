import { mount } from '@vue/test-utils'
import Container from '../Container.vue'
import ElHeader from '../components/header'
import ElFooter from '../components/footer'
import ElMain from '../components/main'
import ElAside from '../components/aside'
describe('Container.vue', () => {
  it('Container snapshot', () => {
    const wrapper = mount(Container)
    expect(wrapper.element).toMatchSnapshot()
  })
  describe('Container class', () => {
    it('contain has class .el-container', () => {
      const wrapper = mount(Container)
      expect(wrapper.find('.el-container').exists()).toBe(true)
    })
    it ('contian has class .is-vertical by prop direction=vertical', () => {
      const wrapper = mount(Container, {
        props: {
          direction: 'vertical',
        },
      })
      expect(wrapper.find('.is-vertical').exists()).toBe(true)
    })
    it ('contian has class .is-vertical by prop direction=horizontal', () => {
      const wrapper = mount(Container, {
        props: {
          direction: 'horizontal',
        },
      })
      expect(wrapper.find('.is-vertical').exists()).toBe(false)
    })
  })
  describe('Container slot', () => {
    it('slot has el-header', () => {
      const wrapper = mount({
        template: `
          <el-container>
            <el-header></el-header>
            <el-main>
              <el-aside></el-aside>
            </el-main>
          </el-container>
        `,
        components: {
          'el-container': Container,
          'el-header': ElHeader,
          'el-main': ElMain,
          'el-aside': ElAside,
        },
      })

      // expect(wrapper.classes()).toBe(true);
      // expect(wrapper.find('.is-vertical').exists()).toBe(true)
      expect(wrapper.find('.is-vertical').exists()).toBe(true)
    })
    it('slot has el-footer', () => {
      const wrapper = mount({
        template: `
          <el-container>
            <el-main>
              <el-aside></el-aside>
            </el-main>
            <el-footer></el-footer>
          </el-container>
        `,
        components: {
          'el-container': Container,
          'el-main': ElMain,
          'el-aside': ElAside,
          'el-footer': ElFooter,
        },
      })

      expect(wrapper.find('.is-vertical').exists()).toBe(true)
    })
  })
})

