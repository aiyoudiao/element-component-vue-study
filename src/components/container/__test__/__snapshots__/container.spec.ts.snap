// Vitest Snapshot v1

exports[`Container.vue > Container snapshot 1`] = `
<section
  class="el-container"
>
  
  
</section>
`;
